### PR TITLE
p256: Add signing & encryption abstraction w/ key derivation

### PR DESCRIPTION
**Pull request by emostov**
- Add p256 signing
- Remove use of EphemeralKey for ECDH and instead use SecretKey. EphemeralKey does not have APIs to be persisted. While using EphemeralKey is the recommended usage pattern from the p256 crate authors, we want anyone to be able to encrypt to a quorum key without any interaction. Please advise if this is an immediate security issue.
- Add to/from bytes for all keys
- Add to to/from hex file for all keys
- Introduce top level P256 abstraction that supports both singing and encryption. This abstraction uses master secret from which we derive separate singing and encryption secret keys.

### CODE REVIEW

File: src/qos_p256/src/lib.rs [hunk 1]
```diff
@@ -28,4 +42,278 @@ pub enum P256Error {
...
+	seed: [u8; 64],
+	derive_path: &[u8],
+) -> Result<[u8; 32], P256Error> {
+	let hk = Hkdf::<Sha512>::new(Some(derive_path), &seed);
+
+	let mut buf = [0u8; P256_SECRET_LEN];
+	hk.expand(&[], &mut buf).map_err(|_| P256Error::HkdfExpansionFailed)?;
+
+	Ok(buf)
+}
```
Comment (emostov) [inline src/qos_p256/src/lib.rs]: Does this look right for doing derivation from a master seed for p256 keys?

File: src/qos_p256/src/encrypt.rs [hunk 2]
```diff
@@ -332,4 +356,33 @@ mod tests {
...
+		let alice_public2 =
+			P256EncryptPublic::from_bytes(&public_key_bytes).unwrap();
+
+		let plaintext = b"rust test message";
+
+		let serialized_envelope = alice_public2.encrypt(plaintext).unwrap();
+
+		let decrypted = alice_pair.decrypt(&serialized_envelope).unwrap();
+
+		assert_eq!(decrypted, plaintext);
```
Comment (jack-kearney) [inline src/qos_p256/src/encrypt.rs:377]: Kind of redundant, but maybe worth asserting that `alice_public` and `alice_public2` are equivalent? (Essentially just replicating the test case below)

File: src/qos_p256/src/lib.rs [hunk 3]
```diff
@@ -27,4 +49,391 @@ pub enum P256Error {
...
+		Self::QosHex(err)
+	}
+}
+
+/// Helper function to derive a secret from a master seed.
+fn derive_secret(
+	seed: [u8; MASTER_SEED_LEN],
+	derive_path: &[u8],
+) -> Result<[u8; 32], P256Error> {
+	let hk = Hkdf::<Sha512>::new(Some(derive_path), &seed);
```
Comment (mikelodder7) [inline src/qos_p256/src/lib.rs]: Hkdf is overkill but still secure

File: src/qos_p256/src/lib.rs [hunk 4]
```diff
@@ -27,4 +49,391 @@ pub enum P256Error {
...
+	Ok(buf)
+}
+
+/// Helper function to generate a `N` length byte buffer.
+fn non_zero_bytes_os_rng<const N: usize>() -> [u8; N] {
+	loop {
+		let mut key = [0u8; N];
+		OsRng.fill_bytes(&mut key);
+
+		if key.iter().all(|bit| *bit == 0) {
```
Comment (mikelodder7) [inline src/qos_p256/src/lib.rs:100]: The odds of this happening is incredibly low that most libraries don’t even bother. It’s fine that you do it but probably not necessary

File: src/qos_p256/src/lib.rs [hunk 5]
```diff
@@ -27,4 +49,391 @@ pub enum P256Error {
...
+	fn from(err: qos_hex::HexError) -> Self {
+		Self::QosHex(err)
+	}
+}
+
+/// Helper function to derive a secret from a master seed.
+fn derive_secret(
+	seed: [u8; MASTER_SEED_LEN],
+	derive_path: &[u8],
+) -> Result<[u8; 32], P256Error> {
```
Comment (mikelodder7) [inline src/qos_p256/src/lib.rs]: You’ll want 64 bytes to be returned due to bias in the keys. There should be a from_wide_bytes to handle this. This eliminates any bias due to the reduction with mod q

File: src/qos_p256/src/lib.rs [hunk 6]
```diff
@@ -27,4 +49,391 @@ pub enum P256Error {
...
+
+	/// Read the raw, hex encoded master from a file.
+	// TODO: implement utils that go to/from bytes so we can avoid string
+	// serialization. https://github.com/tkhq/qos/issues/153.
+	pub fn from_hex_file<P: AsRef<Path>>(path: P) -> Result<Self, P256Error> {
+		let hex_bytes = std::fs::read(path).map_err(|e| {
+			P256Error::IOError(format!("failed to read master seed: {}", e))
+		})?;
+
+		let hex_string = String::from_utf8(hex_bytes)
```
Comment (mikelodder7) [inline src/qos_p256/src/lib.rs:195]: I think the `hex` can decode from bytes directly removing the need to convert to a string first

File: src/qos_p256/src/lib.rs [hunk 7]
```diff
@@ -27,4 +49,391 @@ pub enum P256Error {
...
+			// try again if we got all zeros
+		} else {
+			return key;
+		}
+	}
+}
+
+/// P256 private key pair for signing and encryption. Internally this uses a
+/// separate secret for signing and encryption.
+pub struct P256Pair {
```
Comment (mikelodder7) [inline src/qos_p256/src/lib.rs:111]: I would derive the `zeroize` to clear the memory on drop for the seed and keys

File: src/qos_p256/src/lib.rs [hunk 8]
```diff
@@ -27,4 +49,391 @@ pub enum P256Error {
...
+	#[test]
+	fn signatures_are_deterministic() {
+		let message = b"a message to authenticate";
+
+		let pair = P256Pair::generate().unwrap();
+		(0..100)
+			.map(|_| pair.sign(message).unwrap())
+			.collect::<Vec<_>>()
+			.windows(2)
+			.for_each(|slice| assert_eq!(slice[0], slice[1]));
```
Comment (jack-kearney) [inline src/qos_p256/src/lib.rs:302]: EXTREME DETERMINISM

File: src/qos_p256/src/lib.rs [hunk 9]
```diff
@@ -27,4 +49,391 @@ pub enum P256Error {
...
+
+		let plaintext = b"rust test message";
+
+		let serialized_envelope = alice_public.encrypt(plaintext).unwrap();
+
+		let bob_pair = P256Pair::generate().unwrap();
+
+		assert_eq!(
+			bob_pair.decrypt(&serialized_envelope).unwrap_err(),
+			P256Error::AesGcm256DecryptError
```
Comment (jack-kearney) [inline src/qos_p256/src/lib.rs:356]: Nit: Feels like this should be a higher level error like `P256Error::DecryptionError`

File: src/qos_p256/src/lib.rs [hunk 10]
```diff
@@ -27,4 +49,391 @@ pub enum P256Error {
...
+	pub fn public_key(&self) -> P256Public {
+		P256Public {
+			encrypt_public: self.encrypt_private.public_key(),
+			sign_public: self.sign_private.public_key(),
+		}
+	}
+
+	/// Create `Self` from a master seed.
+	pub fn from_master_seed(
+		master_seed: [u8; MASTER_SEED_LEN],
```
Comment (mikelodder7) [inline src/qos_p256/src/lib.rs]:
```suggestion
		master_seed: &[u8; MASTER_SEED_LEN],
```

Comment (mikelodder7) [reply src/qos_p256/src/lib.rs]: You don’t want the master seed copied around in memory so use a reference

File: src/qos_p256/src/sign.rs [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+	#[must_use]
+	pub fn generate() -> Self {
+		Self { private: SigningKey::random(&mut OsRng) }
+	}
+
+	/// Sign the message and return the. Signs the SHA512 digest of
+	/// the message.
+	pub fn sign(&self, message: &[u8]) -> Result<Vec<u8>, P256Error> {
+		let digest = sha2::Sha512::digest(message);
+		let signature: Signature = self.private.sign(&digest);
```
Comment (mikelodder7) [inline src/qos_p256/src/sign.rs]: I think sign is already hashing the message so this is a double hash which is unnecessary

File: src/qos_p256/src/sign.rs [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+	/// Verify a `signature` and `message` against this private key. Verifies
+	/// the SHA512 digest of the message.
+	///
+	/// Returns Ok if the signature is good.
+	pub fn verify(
+		&self,
+		message: &[u8],
+		signature: &[u8],
+	) -> Result<(), P256Error> {
+		let digest = sha2::Sha512::digest(message);
```
Comment (mikelodder7) [inline src/qos_p256/src/sign.rs]: Same as above

File: src/qos_p256/src/lib.rs (hunk 10 above)
Comment (emostov) [reply src/qos_p256/src/lib.rs]: I assume use a reference with a named lifetime in the struct as well?

Comment (mikelodder7) [reply src/qos_p256/src/lib.rs]: No that’s not necessary. The struct can own the seed. What you don’t want is an accidental third copy in ram after this call. The reference prevents that

File: src/qos_p256/src/lib.rs (hunk 4 above)
Comment (emostov) [reply src/qos_p256/src/lib.rs:100]: good to know. Don't expect key gen to be a bottle neck for now so will leave in

File: src/qos_p256/src/lib.rs (hunk 5 above)
Comment (emostov) [reply src/qos_p256/src/lib.rs]: Struggling to find an api for the RustCrypto secret key types that take something that isn't the exact size. Mainly looking here https://github.com/RustCrypto/traits/blob/master/elliptic-curve/src/secret_key.rs which is what the `SigningKey` encapsulates and seems to the lowest level secret abstraction aware of EC stuff

File: src/qos_p256/src/lib.rs (hunk 10 above)
Comment (emostov) [reply src/qos_p256/src/lib.rs]: ea831facae3d9e96e49be63254a13262557abe5c

File: src/qos_p256/src/sign.rs (hunk 11 above)
Comment (emostov) [reply src/qos_p256/src/sign.rs]: Yup, confirmed: https://github.com/RustCrypto/signatures/blob/ace98c82486a322c2132345e1ec5736f3d0e3041/ecdsa/src/sign.rs#L281

File: src/qos_p256/src/sign.rs (hunk 12 above)
Comment (emostov) [reply src/qos_p256/src/sign.rs]: 82f34cda0fef5825399ffcbe2af17b7fe99eb561

File: src/qos_p256/src/sign.rs (hunk 11 above)
Comment (emostov) [reply src/qos_p256/src/sign.rs]: 82f34cda0fef5825399ffcbe2af17b7fe99eb561

File: src/qos_p256/src/lib.rs (hunk 7 above)
Comment (emostov) [reply src/qos_p256/src/lib.rs:111]: c22a8e28c253e94645dcfb617a130cd1bff70d90

File: src/qos_p256/src/lib.rs (hunk 6 above)
Comment (emostov) [reply src/qos_p256/src/lib.rs:195]: Ya, it does. But I think we can implement it ourselves in the not to far future so would save us at least one dep :)

File: src/qos_p256/src/encrypt.rs (hunk 2 above)
Comment (emostov) [reply src/qos_p256/src/encrypt.rs:377]: We don't have `PartialEq` derived on those structs so we can't compare them for equality

File: src/qos_p256/src/lib.rs (hunk 3 above)
Comment (emostov) [reply src/qos_p256/src/lib.rs]: @ace0 any thoughts here?

**Pull request merged**